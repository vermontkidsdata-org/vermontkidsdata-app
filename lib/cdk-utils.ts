import { Duration } from "aws-cdk-lib";
import { AuthorizationType, IResource, LambdaIntegration, MethodOptions, MethodResponse, RequestAuthorizer, Resource } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { Role } from "aws-cdk-lib/aws-iam";
import { Runtime, Tracing } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import { join } from "path";

export interface AuthInfo {
  authorizationType: AuthorizationType,
  authorizer: RequestAuthorizer,
  methodResponses: MethodResponse[],
}

export function addResource(root: Resource, path: string): IResource {
  const parts = path.split('/').filter(p => p);
  let resource: IResource = root;
  for (const part of parts) {
    const existingChild = resource.getResource(part);
    if (existingChild) {
      resource = existingChild;
      continue;
    }
    resource = resource.addResource(part);
  }
  return resource;
}

export type OnAddCallback = (fn: NodejsFunction) => void;

export function makeLambda(props: {
  scope: Construct,
  name: string,
  entry: string,
  handler?: string,
  timeout?: Duration,
  commonEnv: Record<string, string>,
  onAdd?: OnAddCallback,
  serviceTable?: ITable,
  role?: Role,
}): NodejsFunction {
  const { scope, entry: _entry, handler: _handler, commonEnv, onAdd, name, timeout, serviceTable, role } = props;
  const entry = /[^0-9a-zA-Z]/.test(_entry) ? join(__dirname, '../src', _entry) : _entry;
  const handler = _handler || 'handler';

  const fn = new NodejsFunction(scope, name, {
    entry,
    handler: handler,
    memorySize: 1024,
    runtime: Runtime.NODEJS_20_X,
    timeout: timeout ?? Duration.seconds(29),
    environment: commonEnv,
    role,
    tracing: Tracing.ACTIVE,
  });

  if (onAdd) {
    onAdd(fn);
  }

  if (serviceTable) {
    serviceTable.grantReadData(fn);
  }

  return fn;
}

export function addLambdaResource(props: {
  scope: Construct,
  root: Resource,
  method: string,
  path: string,
  entry: string,
  commonEnv: Record<string, string>,
  onAdd?: OnAddCallback,
  methodOptions: MethodOptions,
  serviceTable?: ITable,
  role?: Role,
}): {
  fn: NodejsFunction,
} {
  const { scope, root, commonEnv, onAdd, methodOptions, method, path, entry, serviceTable, role } = props;

  const fn = makeLambda({
    scope,
    name: `${method} ${path}`,
    entry,
    commonEnv,
    onAdd,
    serviceTable,
    role,
  });

  const postCompletion = addResource(root, path);
  postCompletion.addMethod(method, new LambdaIntegration(fn), methodOptions);

  return { fn };
}