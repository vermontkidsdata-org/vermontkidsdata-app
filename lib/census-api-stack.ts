import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3Deploy from 'aws-cdk-lib/aws-s3-deployment';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as path from 'path';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';

export class CensusAPIStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
      super(scope, id, props);

      const citysdkLayer = new lambda.LayerVersion(this, 'CitySDK Layer', {
        code: lambda.Code.fromAsset('src/layers/citysdk-utils'),
        compatibleRuntimes: [ lambda.Runtime.NODEJS_12_X, lambda.Runtime.NODEJS_14_X ]
      });

      const helloWorld = new NodejsFunction(this, 'hello-world-function', {
        memorySize: 128,
        timeout: cdk.Duration.seconds(5),
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'main',
        entry: path.join(__dirname, `/../src/hello.ts`)
      });

      // Package up the templates to S3
      const templateBucket = new s3.Bucket(this, "Template Distribution Bucket", {});

      const deployment = new s3Deploy.BucketDeployment(this, "Render templates", {
        destinationBucket: templateBucket,
        sources: [s3Deploy.Source.asset(path.join(__dirname, "../views"))]
      });

      const apiChartBarFunction = new NodejsFunction(this, 'Bar Chart API Function', {
        memorySize: 128,
        timeout: cdk.Duration.seconds(30),
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'bar',
        entry: path.join(__dirname, `/../src/chartsRender.ts`)
      });

      const renderChartBarFunction = new NodejsFunction(this, 'Bar Chart Render Function', {
        memorySize: 128,
        timeout: cdk.Duration.seconds(30),
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'bar',
        entry: path.join(__dirname, `/../src/chartsApi.ts`),
        environment: {
          templateBucket: templateBucket.s3UrlForObject()
        }
      });

      const testCensusFunction = new NodejsFunction(this, 'Test Census Function', {
        memorySize: 1024,
        timeout: cdk.Duration.seconds(15),
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'getCensus',
        entry: path.join(__dirname, `/../src/testcitylambda.ts`),
        bundling: {
          minify: true,
          externalModules: [ 'citysdk' ]
        },
        layers: [ citysdkLayer ]
      });

      const testDBFunction = new NodejsFunction(this, 'Test DB Function', {
        memorySize: 1024,
        timeout: cdk.Duration.seconds(15),
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'queryDB',
        entry: path.join(__dirname, `/../src/testcitylambda.ts`),
        bundling: {
          minify: true,
          externalModules: [ 'citysdk' ]
        },
        layers: [ citysdkLayer ]
      });

      testDBFunction.addToRolePolicy(new iam.PolicyStatement({
        actions: ["secretsmanager:GetSecretValue"],
        resources: ["*"]
      }));

      const api = new RestApi(this, "Census");
      const hello = api.root.addResource("hello");
      hello.addMethod("GET", new LambdaIntegration(helloWorld));

      const rRender = api.root.addResource("render");
      const rRenderChart = rRender.addResource("chart");
      const rRenderChartBar = rRenderChart.addResource("bar");
      rRenderChartBar.addResource("{chartId}").addMethod("GET", new LambdaIntegration(renderChartBarFunction));

      const rChart = api.root.addResource("chart");
      const rChartBar = rChart.addResource("bar");
      rChartBar.addResource("{chartId}").addMethod("GET", new LambdaIntegration(apiChartBarFunction));

      const testCensusResource = api.root.addResource("census");
      testCensusResource.addMethod("GET", new LambdaIntegration(testCensusFunction));

      const testDBResource = api.root.addResource("testdb");
      testDBResource.addMethod("GET", new LambdaIntegration(testDBFunction));
    }
}