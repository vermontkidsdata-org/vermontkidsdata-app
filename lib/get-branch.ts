import { exec } from 'child_process';

export function getBranch(callback: (branch: string) => void): void {
    exec('git rev-parse --abbrev-ref HEAD', (err, stdout, stderr) => {
        if (err) {
            if (err.message.indexOf('not a git repository') >= 0) {
                callback('main');        // Pretend it's just main
            } else {
                console.error(`getBranch Error: <<${err}>>`);
                process.exit(1);
            }
        } else if (typeof stdout === 'string') {
            callback(stdout.trim());
        } else {
            console.error(`getBranch unknown response type`);
            process.exit(1);
        }
    });
}