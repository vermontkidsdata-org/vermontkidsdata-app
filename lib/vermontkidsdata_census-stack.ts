import { SecretValue, Stack, StackProps } from 'aws-cdk-lib';
import { BuildSpec, EventAction, FilterGroup, GitHubSourceCredentials, Project, Source } from 'aws-cdk-lib/aws-codebuild';
import { CodeBuildStep, CodePipeline, CodePipelineSource, ShellStep } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';
import { PipelineDevStage } from './pipeline-dev-stage';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as subs from 'aws-cdk-lib/aws-sns-subscriptions';
import * as notify from 'aws-cdk-lib/aws-codestarnotifications';

export class VermontkidsdataCensusStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const owner = 'vermontkidsdata-org';
    const repo = 'vermontkidsdata_census';
    const branch = 'master';
    const region = 'us-east-1';
    const account = '439348011602';
    const gitHubTokenSecret = "VKDPipelineGitHubToken-ZPZw3Z";

    const secretArn = `arn:aws:secretsmanager:${region}:${account}:secret:${gitHubTokenSecret}`;

    new GitHubSourceCredentials(this, 'GitHubCreds', {
      accessToken: SecretValue.secretsManager(`arn:aws:secretsmanager:${region}:${account}:secret:${gitHubTokenSecret}`, {
        jsonField: 'access-token',
      }),
    });
    
    const pipelineSpec = BuildSpec.fromObject({
      version: 0.2,
      phases: {
          install: {
              commands: ['n latest', 'node -v', 'npm ci', 'npm ci --prefix src/layers/citysdk-utils/nodejs'],
          },
          build: {
              commands: ['npx cdk synth']
          }
      }
  });
  
  const synthAction = new CodeBuildStep(`Synth`, {
      input: CodePipelineSource.gitHub(`${owner}/${repo}`, branch, {
          authentication: SecretValue.secretsManager(secretArn, {
              jsonField: 'access-token',
          }),
      }),
      partialBuildSpec: pipelineSpec,
      commands: [],
  });

  const pipeline = new CodePipeline(this, `Pipeline`, {
      synth: synthAction,
      dockerEnabledForSynth: true,
      // need this if you're actually deploying to multiple accounts
      // crossAccountKeys: true,
  });

  const source = Source.gitHub({
      owner: owner,
      repo: repo,
      webhook: true,
      webhookFilters: [
        FilterGroup.inEventOf(
          EventAction.PULL_REQUEST_CREATED,
          EventAction.PULL_REQUEST_UPDATED,
          EventAction.PULL_REQUEST_REOPENED,
        ).andBranchIsNot('master'),
      ],
      reportBuildStatus: true,
    });
    
    const prSpec = BuildSpec.fromObject({
      version: 0.2,
      phases: {
        install: {
          commands: ['n latest', 'node -v', 'npm ci'],
        },
        build: {
          commands: ['npx cdk synth', 'npm run test']
        }
      }
    });
    
    new Project(this, 'PullRequestProject', {
      source,
      buildSpec: prSpec,
      concurrentBuildLimit: 1,
    });
    
    pipeline.addStage(new PipelineDevStage(this, "PipelineDevStage", {
      env: { account: "439348011602", region: "us-east-1" },
    }));

    // SNS topic for notification
    const topic = new sns.Topic(this, "Topic", {
      displayName: "Pipeline build results"
    });
    topic.addSubscription(new subs.EmailSubscription("gbisaga@gmail.com"));

    topic.addToResourcePolicy(new iam.PolicyStatement({
      principals: [ new iam.ServicePrincipal("codestar-notifications.amazonaws.com") ],
      actions: [ "SNS:Publish" ],
      resources: [ topic.topicArn ]
    }))

    // Have to do this to ensure pipeline is built before the notification added - but we
    // can't make any more modifications after this.
    pipeline.buildPipeline();


    // const notification = new notify.NotificationRule(this, "Pipeline Notification Rule", {
    //   events: [ 'codepipeline-pipeline-stage-execution-succeeded', 'codepipeline-pipeline-stage-execution-failed' ],
    //   source: pipeline.pipeline,
    //   targets: [ topic ],
    // });

  }
}
