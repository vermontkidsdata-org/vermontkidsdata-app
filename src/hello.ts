import {APIGatewayProxyEventV2, APIGatewayProxyResultV2} from 'aws-lambda';

export async function main(
  event: APIGatewayProxyEventV2,
): Promise<APIGatewayProxyResultV2> {
  console.log('event ðŸ‘‰', event);
I am an error!
  return {
    body: JSON.stringify({message: 'Successful lambda invocation for Dave'}),
    statusCode: 200,
  };
}
