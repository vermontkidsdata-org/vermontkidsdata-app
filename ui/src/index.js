import React from 'react';
import ReactDOM from 'react-dom';
import { Authenticator, useAuthenticator } from '@aws-amplify/ui-react';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';
import Amplify from 'aws-amplify';
// import config from './aws-exports';
import { Link, BrowserRouter as Router } from 'react-router-dom';
// Amplify.configure(config);

ReactDOM.render(

    <Router>
          <App />
    </Router>
  ,
  document.getElementById('root')
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();

/*
<React.StrictMode>
      <Authenticator.Provider>
          <App />
      </Authenticator.Provider>
  </React.StrictMode>
 */
