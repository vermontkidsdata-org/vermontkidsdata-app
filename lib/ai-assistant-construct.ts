import { Duration } from "aws-cdk-lib";
import { MethodOptions, RestApi } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { ISecret, Secret } from "aws-cdk-lib/aws-secretsmanager";
import { StateMachine } from "aws-cdk-lib/aws-stepfunctions";
import { LambdaInvoke } from "aws-cdk-lib/aws-stepfunctions-tasks";
import { Construct } from "constructs";
import { IN_PROGRESS_ERROR } from "../src/ai-utils";
import { AuthInfo, OnAddCallback, addLambdaResource, makeLambda } from "./cdk-utils";

interface AIAssistantProps {
  api: RestApi;
  methodOptions?: MethodOptions;
  auth?: AuthInfo;
  commonEnv: Record<string, string>;
  onAdd?: OnAddCallback;
  serviceTable: ITable;
  secret: ISecret;
}
const VKD_API_KEY = '09848734-8745-afrt-8745-8745873487';

export class AIAssistantConstruct extends Construct {
  _props: AIAssistantProps;

  constructor(scope: Construct, id: string, props: AIAssistantProps) {
    super(scope, id);
    this._props = props;

    const { api, commonEnv, onAdd, methodOptions, auth, serviceTable, secret, } = props;
    const aiAssistantRoot = api.root.addResource('ai');

    const aiSecret = Secret.fromSecretNameV2(this, 'AI Secret', 'openai-config');

    const aiCommonEnv = {
      ...commonEnv,
      // ASSISTANT_ID: 'asst_TGWe7LGoOMxyTuRa9uOdLXwD',
      ASSISTANT_ID: 'asst_nJKMeBh1KxrqsrL9jGheMcHX',
      VKD_STATE_MACHINE_ARN: '',
      SERVICE_TABLE: serviceTable.tableName,
      AI_SECRET_NAME: aiSecret.secretName,
      VKD_API_KEY,
    };

    const startOpenAICompletion = makeLambda({
      scope: this,
      name: 'Start OpenAI Completion',
      entry: 'ai-start-openai-completion.ts',
      timeout: Duration.minutes(15),
      commonEnv: aiCommonEnv,
      onAdd,
    });
    ((fn) => {
      serviceTable.grantReadWriteData(fn);
      aiSecret.grantRead(fn);
      secret.grantRead(fn);
    })(startOpenAICompletion);

    const checkOpenAICompletion = makeLambda({
      scope: this,
      name: 'Check OpenAI Completion',
      entry: 'ai-check-openai-completion.ts',
      commonEnv: aiCommonEnv,
      onAdd,
    });
    ((fn) => {
      serviceTable.grantReadWriteData(fn);
      aiSecret.grantRead(fn);
      secret.grantRead(fn);
    })(checkOpenAICompletion);

    const checkComplete = new LambdaInvoke(this, 'Invoke Check OpenAI Completion', {
      lambdaFunction: checkOpenAICompletion,
      outputPath: '$.Payload',
    });
    checkComplete.addRetry({
      errors: [IN_PROGRESS_ERROR],
      interval: Duration.seconds(5),
      maxAttempts: 120,
      backoffRate: 1.0,
    });

    const definition =
      new LambdaInvoke(this, 'Invoke Start OpenAI Completion', {
        lambdaFunction: startOpenAICompletion,
        outputPath: '$.Payload',
      }).next(checkComplete);

    const sf = new StateMachine(this, 'AIAssistantStateMachine', {
      tracingEnabled: true,
      definition,
    });
    startOpenAICompletion.grantInvoke(sf);
    checkOpenAICompletion.grantInvoke(sf);

    aiCommonEnv.VKD_STATE_MACHINE_ARN = sf.stateMachineArn;

    // Mix in the auth (POST endpoints only)
    const methodOptionsWithAuth = {
      ...methodOptions,
      // ...auth,
    };

    (({ fn }) => {
      sf.grantStartExecution(fn);
      aiSecret.grantRead(fn);
    })(addLambdaResource({
      scope: this,
      root: aiAssistantRoot,
      method: 'POST',
      path: 'completion',
      entry: 'ai-post-completion.ts',
      commonEnv: aiCommonEnv,
      onAdd,
      methodOptions: methodOptionsWithAuth,
    }));

    (({ fn }) => {
      serviceTable.grantReadData(fn);
    })(addLambdaResource({
      scope: this,
      root: aiAssistantRoot,
      method: 'GET',
      path: 'completion/{id}/{sortKey}',
      entry: 'ai-get-completion.ts',
      commonEnv: aiCommonEnv,
      onAdd,
      methodOptions: methodOptionsWithAuth,
    }));
  }
}

