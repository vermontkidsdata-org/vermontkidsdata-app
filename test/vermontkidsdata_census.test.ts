import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CensusAPIStack } from '../lib/census-api-stack';
import * as hello from '../src/hello';

// import { Template } from 'aws-cdk-lib/assertions';
import * as VermontkidsdataCensus from '../lib/vermontkidsdata_census-stack';
import { APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from 'aws-lambda';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/vermontkidsdata_census-stack.ts
describe('Lambda', () => {
    it('Creates Lambda', () => {
        const app = new cdk.App();
        // WHEN
        const lambdaStack = new CensusAPIStack(app, 'CensusAPIDevStack');
        // THEN
        const template = Template.fromStack(lambdaStack);

        template.hasResourceProperties('AWS::Lambda::Function', {
            Handler: "index.main"
        });
    });
    
    it('Does the snapshot thing', () => {
        const app = new cdk.App();
        const lambdaStack = new CensusAPIStack(app, 'CensusAPIDevStack');
        const template = Template.fromStack(lambdaStack);

        /*
        * const stack = new Stack(app, 'MyStack');
        * const cfn = SynthUtils.toCloudFormation(stack);
        * expect(cfn).toMatchSnapshot(getMatchObject(cfn.Resources));
        */
        expect(template.toJSON()).toMatchSnapshot("CensusAPIDevStack-basic");
    });

});

describe('Hello code', () => {
    it('Returns invocation message', async () => {
        const event: APIGatewayProxyEventV2 = {
            headers: {

            },
            isBase64Encoded: false,
            rawPath: '',
            rawQueryString: '',
            requestContext: {} as any,
            routeKey: '',
            version: ''
        };
        const response: any = await hello.main(event);
        const body = JSON.parse(response['body']);
        expect(body.message).toMatch(/^Successful lambda invocation.*/);
    });
});
