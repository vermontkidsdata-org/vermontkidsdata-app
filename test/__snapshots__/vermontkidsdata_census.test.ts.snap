// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda Does the snapshot thing: CensusAPIDevStack-basic 1`] = `
Object {
  "Outputs": Object {
    "CensusEndpoint2A3F9C45": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "Census06374A65",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "CensusDeploymentStageprodAD5A6255",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Census06374A65": Object {
      "Properties": Object {
        "Name": "Census",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "CensusAccountA10763DC": Object {
      "DependsOn": Array [
        "Census06374A65",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "CensusCloudWatchRole61DFD17A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "CensusCloudWatchRole61DFD17A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CensusDeploymentC9A3DCD88b9fd2bf0f275b88e42c19a81dd9633c": Object {
      "DependsOn": Array [
        "CensuscensusGETD029A41B",
        "CensuscensusBAE55230",
        "CensushelloGET16A02370",
        "Censushello44D10716",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "CensusDeploymentStageprodAD5A6255": Object {
      "DependsOn": Array [
        "CensusAccountA10763DC",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "CensusDeploymentC9A3DCD88b9fd2bf0f275b88e42c19a81dd9633c",
        },
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "CensuscensusBAE55230": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "Census06374A65",
            "RootResourceId",
          ],
        },
        "PathPart": "census",
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "CensuscensusGETApiPermissionCensusAPIDevStackCensusA47A87F6GETcensus702E9E21": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TestCensusFunctionB81B1FE4",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "Census06374A65",
              },
              "/",
              Object {
                "Ref": "CensusDeploymentStageprodAD5A6255",
              },
              "/GET/census",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CensuscensusGETApiPermissionTestCensusAPIDevStackCensusA47A87F6GETcensus46AE132C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TestCensusFunctionB81B1FE4",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "Census06374A65",
              },
              "/test-invoke-stage/GET/census",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CensuscensusGETD029A41B": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "TestCensusFunctionB81B1FE4",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "CensuscensusBAE55230",
        },
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "Censushello44D10716": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "Census06374A65",
            "RootResourceId",
          ],
        },
        "PathPart": "hello",
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "CensushelloGET16A02370": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "helloworldfunction8252912C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "Censushello44D10716",
        },
        "RestApiId": Object {
          "Ref": "Census06374A65",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CensushelloGETApiPermissionCensusAPIDevStackCensusA47A87F6GEThello22053947": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "helloworldfunction8252912C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "Census06374A65",
              },
              "/",
              Object {
                "Ref": "CensusDeploymentStageprodAD5A6255",
              },
              "/GET/hello",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CensushelloGETApiPermissionTestCensusAPIDevStackCensusA47A87F6GEThello4E32F076": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "helloworldfunction8252912C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "Census06374A65",
              },
              "/test-invoke-stage/GET/hello",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CitySDKLayer253B626B": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs12.x",
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9cf9ae33b27adc8341fa5f521f5061ab51a8d610e2116434657a0bd8411b61e8.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "TestCensusFunctionB81B1FE4": Object {
      "DependsOn": Array [
        "TestCensusFunctionServiceRoleC9BF2E43",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e743fa111a19a1fab0d4b2a0970e98f47412b323ba70431ea33cc1c223ed784d.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.main",
        "Layers": Array [
          Object {
            "Ref": "CitySDKLayer253B626B",
          },
        ],
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCensusFunctionServiceRoleC9BF2E43",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCensusFunctionServiceRoleC9BF2E43": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "helloworldfunction8252912C": Object {
      "DependsOn": Array [
        "helloworldfunctionServiceRole0327FCC6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "67334357cd05e5a72e18f90fe2ce64429a4a8e0259794c70b9d5aac574050690.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.main",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "helloworldfunctionServiceRole0327FCC6",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "helloworldfunctionServiceRole0327FCC6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
